src/tuwien/auto/calimero/knxnetip/util/SRP.java

```diff
@@ -0,0 +1,220 @@
+
+/**
+ * Search Request Parameter Block (SRP).
+ * <p>
+ * A SRP is used to transfer additional information in a KNXnet/IP extended search requests.
+ * <p>
+ * @author Karsten Heimrich
+ */
+public final class SRP
+{
+	/**
+	 * The invalid code is used to test the behavior of the KNXnet/IP router or server for unknown
+	 * SRPs. Please do not use in production code.
+	 */
+	public static final int INVALID_SRP = 0x00;
+
+	/**
+	 * The select by programming mode type-code is used to indicate that a client is interested only in the
+	 * response from KNXnet/IP router or server where programming mode is currently enabled.
+	 */
+	public static final int SELECT_BY_PROGRAMMING_MODE = 0x01;
+
+	/**
+	 * The select by MAC address type-code is used to indicate that a client is interested only in the
+	 * response from KNXnet/IP router or server with the given MAC address.
+	 */
+	public static final int SELECT_BY_MAC_ADDRESS = 0x02;
+
+	/**
+	 * The select by service type-code is used to indicate that a client is interested only in the response
+	 * from KNXnet/IP router or server supporting the given service family and in at least the given version.
+	 */
+	public static final int SELECT_BY_SERVICE = 0x03;
+
+	/**
+	 * The request DIBs type-code is used to indicate that a client is interested in at least the given
+	 * description types.
+	 */
+	public static final int REQUEST_DIBS = 0x04;
+
+	private final int size;
+	private final boolean mandatory;
+	private final int type;
+	private final byte[] data;
+
+	private static final int SrpHeaderSize = 2;
+
+	/**
+	 * Creates a new SRP and initializes basic fields.
+	 * <p>
+	 * @param srpType one of the search request parameter block constants of this class
+	 * @param isMandatory to be evaluated by a KNXnet/IP router or server device
+	 * @param data byte array containing additional data, ignored for select by programming mode SRPs
+	 */
+	public SRP(final int srpType, final boolean isMandatory, final byte... data)
+	{
+		type = srpType;
+		mandatory = isMandatory;
+
+		switch (srpType) {
+		case SELECT_BY_MAC_ADDRESS:
+			if (data.length != 6)
+				throw new KNXIllegalArgumentException("MAC size does not match expected value");
+			size = SrpHeaderSize + 6;
+			this.data = data;
+			break;
+		case SELECT_BY_SERVICE:
+			if (data.length != 2)
+				throw new KNXIllegalArgumentException("service info size does not match expected value");
+			size = SrpHeaderSize + 2;
+			this.data = data;
+			break;
+		case REQUEST_DIBS:
+			if (data.length < 1)
+				throw new KNXIllegalArgumentException("requested DIBs size does not match expected value");
+			if ((data.length % 2) != 0) {
+				size = SrpHeaderSize + (data.length + 1);
+				this.data = new byte[size - SrpHeaderSize];
+				System.arraycopy(data, 0, this.data, 0, data.length);
+			} else {
+				size = SrpHeaderSize + data.length;
+				this.data = data;
+			}
+			break;
+		case SELECT_BY_PROGRAMMING_MODE: // fallthrough
+		default:
```

Comment:
If this stays a public ctor, I guess I would not make it fallthrough

Reply:
Will fix it.